yarn-project/key-store/src/test_key_store.ts

```diff
@@ -296,18 +306,36 @@ export class TestKeyStore implements KeyStore {
     return Promise.resolve(Fr.fromBuffer(publicKeysHashBuffer));
   }
 
-  #getAccountAddressForMasterNullifierPublicKeyHash(masterNullifierPublicKeyHash: Fr): AztecAddress | undefined {
-    for (const [key, value] of this.#keys.entries()) {
-      if (key.endsWith('-npk_m')) {
+  #getMapMetadataForMasterNullifierPublicKeyHash(masterNullifierPublicKeyHash: Fr) {
+    for (const [key, value] of [...this.#keys.entries(), ...this.#rotatedKeys.entries()]) {
+      if (key.includes('-npk_m')) {
         const computedMasterNullifierPublicKeyHash = poseidon2Hash(Point.fromBuffer(value).toFields());
         if (computedMasterNullifierPublicKeyHash.equals(masterNullifierPublicKeyHash)) {
-          // We extract the account address from the map key
-          const accountAddress = key.split('-')[0];
-          return AztecAddress.fromString(accountAddress);
+          return { key, value };
         }
       }
     }
 
-    return undefined;
+    throw new Error(`Master nullifier public key hash ${masterNullifierPublicKeyHash} does not exist.`);
+  }
+
+  public async rotateMasterNullifierKey(account: AztecAddress, newSecretKey: Fq = Fq.random()) {
+    const storedAccounts = await this.getAccounts();
+
+    if (!storedAccounts.some(storedAccount => storedAccount.equals(account))) {
+      throw new Error('Account does not exist');
+    }
+
+    const newPublicKey = derivePublicKeyFromSecretKey(newSecretKey);
+
+    const previouslyRotatedNumber = this.#rotatedNullifierKeyCount(account);
+    await this.#rotatedKeys.set(`${account.toString()}-nsk_m-${previouslyRotatedNumber}`, newSecretKey.toBuffer());
+    await this.#rotatedKeys.set(`${account.toString()}-npk_m-${previouslyRotatedNumber}`, newPublicKey.toBuffer());
+  }
+
+  #rotatedNullifierKeyCount(account: AztecAddress): number {
```

Comment:
It seems to me that it would be simpler to just store all the npk_m pertaining to one account in 1 long buffer uner one key instead of dealing with this.